source/extensions/config_subscription/grpc/grpc_mux_failover.h

```diff
@@ -103,41 +173,144 @@ template <class RequestType, class ResponseType> class GrpcMuxFailover {
+      parent_.connected_to_primary_ = true;
+      parent_.connecting_to_primary_ = false;
       parent_.grpc_mux_callbacks_.onDiscoveryResponse(std::move(message), control_plane_stats);
     }
 
     void onWriteable() override {
-      // TODO(adisuissa): At the moment this is a pass-through method. Once the
-      // implementation matures, this call will be updated.
-      parent_.grpc_mux_callbacks_.onWriteable();
+      if (parent_.connecting_to_primary_ || parent_.connected_to_primary_) {
+        parent_.grpc_mux_callbacks_.onWriteable();
+      }
     }
 
   private:
     GrpcMuxFailover& parent_;
+    uint32_t primary_consecutive_failures_{0};
   };
 
-  // Flags to keep track of the state of connections to primary/failover.
-  // All initialized to false, as there is no connection process during
-  // initialization.
-  bool connecting_to_primary_ : 1;
-  bool connected_to_primary_ : 1;
-  bool connecting_to_failover_ : 1;
-  bool connected_to_failover_ : 1;
+  // A helper class that proxies the callbacks of GrpcStreamCallbacks for the failover service.
+  class FailoverGrpcStreamCallbacks : public GrpcStreamCallbacks<ResponseType> {
+  public:
+    FailoverGrpcStreamCallbacks(GrpcMuxFailover& parent) : parent_(parent) {}
+
+    void onStreamEstablished() override {
+      // Although the failover stream is considered established, need to wait for the
+      // the first response to be received before considering the failover available.
+      // Calling the onStreamEstablished() callback on the GrpcMux object will
+      // trigger the GrpcMux to start sending requests.
+      ASSERT(parent_.connecting_to_failover_ && !parent_.connected_to_failover_ &&
+             !parent_.connecting_to_primary_ && !parent_.connected_to_primary_);
+      parent_.grpc_mux_callbacks_.onStreamEstablished();
+    }
+
+    void onEstablishmentFailure() override {
+      // This will be called when the failover stream fails to establish a connection, or after the
+      // connection was closed.
+      ASSERT((parent_.connecting_to_failover_ || parent_.connected_to_failover_) &&
+             !parent_.connecting_to_primary_ && !parent_.connected_to_primary_);
+      if (!parent_.ever_connected_to_failover_) {
+        ASSERT(parent_.connecting_to_failover_);
+        // If Envoy never established a connecting the failover, it will try to connect to the
+        // primary next.
+        ENVOY_LOG(debug, "Failover xDS stream failed to establish a connection. Attempting to "
+                         "connect to the primary stream.");
+        parent_.connecting_to_failover_ = false;
+        // This will close the stream and prevent the retry timer from
+        // reconnecting to the failover source.
+        parent_.failover_grpc_stream_->closeStream();
+        parent_.grpc_mux_callbacks_.onEstablishmentFailure();
+        parent_.connecting_to_primary_ = true;
+        parent_.primary_grpc_stream_->establishNewStream();
```

Comment:
Should there be some artificial delay added to avoid rapid flip-flopping?

Reply:
Added a retry timer in [26aa991](https://github.com/envoyproxy/envoy/pull/34215/commits/26aa99193d553cb0e8407153157ef927e2b90f7d).
This will eventually be removed once the always trying to connect to the primary will be implemented.